### PR TITLE
.Net PromptTemplateEngine using FunctionRunner.

### PR DESCRIPTION
**Pull request by RogerBarreto**
### Motivation and Context

1. Enable triggering of Hooks when calling functions references in the template. 
#2324 
2. One more step towards Kernel.RunAsync first.

### Description

- Move template engine logic to use `context.FunctionRunner.RunAsync` instead of `function.InvokeAsync`
- Removing responsibility to find functions from the PromptTemplateEngine.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Extensions/TemplateEngine.Basic/Blocks/CodeBlock.cs [hunk 1]
```diff
@@ -101,53 +101,25 @@ public async Task<string> RenderCodeAsync(SKContext context, CancellationToken c
...
-
-        if (!this.GetFunction(context.Functions!, fBlock, out ISKFunction? function))
-        {
-            var errorMsg = $"Function `{fBlock.Content}` not found";
-            this.Logger.LogError(errorMsg);
-            throw new SKException(errorMsg);
-        }
-
-        SKContext contextClone = context.Clone();
+        ContextVariables inputVariables = context.Variables.Clone();
```
Comment (SergeyMenshykh) [inline dotnet/src/Extensions/TemplateEngine.Basic/Blocks/CodeBlock.cs:104]: 1. Please add a comment explaining the reason for the cloning.
2. nit: Consider putting the line into the `else` block to avoid double coning of context variables.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Extensions/Extensions.UnitTests/TemplateEngine/Prompt/PromptTemplateEngineTests.cs [hunk 2]
```diff
@@ -149,16 +151,18 @@ string MyFunctionAsync(SKContext context)
...
         }
 
-        ISKFunction func = SKFunction.FromNativeMethod(Method(MyFunctionAsync), this);
-        Assert.NotNull(func);
+        List<ISKFunction> functions = new()
+        {
+            SKFunction.FromNativeMethod(Method(MyFunctionAsync), this),
+        };
+
+        Assert.NotNull(functions[0]);
```
Comment (dmytrostruk) [inline dotnet/src/Extensions/Extensions.UnitTests/TemplateEngine/Prompt/PromptTemplateEngineTests.cs:159]: Looks like this change is present in a couple of places in tests (replacing one function with list of functions), but we still work only with one function and access it by index `functions[0]`. Is this change required? I think if we revert it, it should still work, but maybe I'm missing something, please let me know.

File: dotnet/src/SemanticKernel.Core/Functions/FunctionCollection.cs [hunk 3]
```diff
@@ -76,6 +76,11 @@ public IFunctionCollection AddFunction(ISKFunction functionInstance)
     /// <inheritdoc/>
     public ISKFunction GetFunction(string pluginName, string functionName)
     {
+        if (string.IsNullOrWhiteSpace(pluginName))
+        {
+            return this.GetFunction(GlobalFunctionsPluginName, functionName);
+        }
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Core/Functions/FunctionCollection.cs:82]: nit: Maybe we can simplify and re-assign `pluginName` instead of calling same function again. In this case we won't increase execution stack and avoid potential infinite recursion (I know it's rare case to happen here, but still :) ).
```suggestion
        pluginName = !string.IsNullOrWhiteSpace(pluginName) ?  pluginName : GlobalFunctionsPluginName;
```

**Review by dmytrostruk — APPROVED** [review]

File: dotnet/src/Extensions/Extensions.UnitTests/TemplateEngine/Prompt/PromptTemplateEngineTests.cs (hunk 2 above)
Comment (RogerBarreto) [reply dotnet/src/Extensions/Extensions.UnitTests/TemplateEngine/Prompt/PromptTemplateEngineTests.cs:159]: The previous implementation breaks when using the Invoke native reflection. 
Says that the target is not the same.

Not sure if this is a .Net bug or something, but when I use within a `List` it works.

@stephentoub do you have any insights what might be the problem here?

Comment (RogerBarreto) [reply dotnet/src/Extensions/Extensions.UnitTests/TemplateEngine/Prompt/PromptTemplateEngineTests.cs:159]: Interestingly enough, I retried with the previous logic and it worked. Not sure what might have happened in my last experiment, seems to be related to local execution. Still not sure what was it but can't replicate it now.

**Review by dmytrostruk — APPROVED** [review]
